### DISCUSSION TITLE
Trouble building Libmesh on HPC system, VectorSetAction is used but never defined.

### DISCUSSION THREAD

**Discussion by TheBEllis** [idaholab/moose#20733] [Q&A Getting Started] [answered]
Hi MOOSE team, I am currently having some issues when trying to get MOOSE installed in my local directory of a HPC system. I am able to build PETSC fine, but when it comes to building libmesh I get a huge array of errors, which are shown in the image below.

![liberr](https://user-images.githubusercontent.com/54721825/162187234-2aae4956-1548-4a50-a309-fd8bffb5fdd0.png)

My currently loaded modules in my HPC environment are cmake-3.19.4-gcc-5-a7pima, git-2.31.0-gcc-5.4.0, python/3.8, gcc/9, openmpi-3.1.3-gcc-7.2.0 , and in my mooseenvironment script I have these variables exported.

CC=mpicc
CXX=mpicxx
F90=mpif90
F77=mpif77
FC=mpif90

This may well be question to ask to people on the HPC side of things, but thought I would ask in case it is something you can help with, if there's any more information I can provide to help just let me know.

**Comment by GiudGiud** [comment, marked as answer]
Hello

Do you have mpich or mvapich available on this machine? We work routinely with those.
The openmpi module is consistent with gcc 9 right? What does `mpicc --version` returns?
For future reference we prefer you copy paste text in triple quotes over screenshots. 

Guillaume

**Reply by TheBEllis** [reply]
Hi, I believe MPICH is available but only up to 3.2. running ```mpicc --version``` returned gcc 7.2.0, I can replace the current module I am running with one that supports gcc 9.2. There are also later gcc versions available, I am currently using the modules I am using just because they are what was recommended. 

I have tried re-running with 
```
./scripts/update_and_rebuild_libmesh.sh --with-mpi --with-cxx-std=2011
```

And I get slightly further in the build I believe, but get this error instead.

```
  CXX      src/numerics/libmesh_opt_la-distributed_vector.lo
In file included from ../src/numerics/distributed_vector.C:21:
./include/libmesh/distributed_vector.h: In member function 'virtual void libMesh::DistributedVector<T>::set(libMesh::numeric_index_type, T)':
./include/libmesh/distributed_vector.h:580:8: error: 'scoped_lock' is not a member of 'std'
  580 |   std::scoped_lock lock(this->_numeric_vector_mutex);
      |        ^~~~~~~~~~~
./include/libmesh/distributed_vector.h:580:8: note: 'std::scoped_lock' is only available from C++17 onwards
./include/libmesh/distributed_vector.h: In member function 'virtual void libMesh::DistributedVector<T>::add(libMesh::numeric_index_type, T)':
./include/libmesh/distributed_vector.h:600:8: error: 'scoped_lock' is not a member of 'std'
  600 |   std::scoped_lock lock(this->_numeric_vector_mutex);
      |        ^~~~~~~~~~~
./include/libmesh/distributed_vector.h:600:8: note: 'std::scoped_lock' is only available from C++17 onwards
make[1]: *** [src/numerics/libmesh_opt_la-distributed_vector.lo] Error 1
make[1]: Leaving directory `/home/ir-elli2/projects/moose/libmesh/build'

```

Would you recommend trying to build with the mpich module despite it being a bit outdated?

**Reply by GiudGiud** [reply]
You need C++17 support so this flag isnt going to work.

A bit outdated is OK, as long as you meet these requirements
https://mooseframework.inl.gov/getting_started/installation/manual_installation_gcc.html

**Reply by GiudGiud** [reply]
mpicc on a different gcc version than your gcc module is a no-go of course.
